src/testnet/run_loop.rs

```diff
@@ -0,0 +1,185 @@
+use super::{Config, Node, BurnchainSimulator};
+
+use std::time;
+use std::thread;
+
+use chainstate::burn::{ConsensusHash};
+
+/// RunLoop is coordinating a simulated burnchain and some simulated nodes
+/// taking turns in producing blocks.
+pub struct RunLoop {
+    config: Config,
+    nodes: Vec<Node>,
+}
+
+impl RunLoop {
+
+    /// Sets up a runloop and nodes, given a config.
+    pub fn new(config: Config) -> Self {
+        // Build a vec of nodes based on config
+        let mut nodes = vec![]; 
+        let mut nodes_confs = config.node_config.clone();
+        for conf in nodes_confs.drain(..) {
+            let node = Node::new(conf, config.burnchain_block_time);
+            nodes.push(node);
+        }
+
+        Self {
+            config,
+            nodes,
+        }
+    }
+
+    /// Starts the testnet runloop.
+    /// 
+    /// This function will block by looping infinitely.
+    /// It will start the burnchain (separate thread), set-up a channel in
+    /// charge of coordinating the new blocks coming from the burnchain and 
+    /// the nodes, taking turns on tenures.  
+    pub fn start(&mut self) {
+
+        // Initialize and start the burnchain.
+        let mut burnchain = BurnchainSimulator::new();
+
+        // Start the burnchain - happening on a separate thread. 
+        // Keep a mpsc::Receiver (burnchain_block_rx), used for receiving blocks, and
+        // a mpsc::Sender, cloned by each node and used for submitting ops (key registrations, 
+        // block commits) on the burnchain.
+        let (burnchain_block_rx, burnchain_op_tx) = burnchain.start(&self.config);
+
+        // Tear-up each node with a mpsc::Sender channeling ops to the burnchain
+        for node in self.nodes.iter_mut() {
+            node.tear_up(burnchain_op_tx.clone());
+        }
+
+        // Wait on the genesis block from the burnchain.
+        let (genesis_block, ops, _) = match burnchain_block_rx.recv() {
+            Ok(res) => res,
+            Err(err) => panic!("Error while expecting genesis block from burnchain: {:?}", err)
+        };
+
+        // Update each node with the genesis block.
+        for node in self.nodes.iter_mut() {
+            node.process_burnchain_block(&genesis_block, &ops);
+        }
+
+        // When producing an anchored block, we need a block from the burnchain, and its parent.
+        // We'll wait on the next block from the burnchain for starting the bootstrap.
+        let (burnchain_block_1, ops, _) = match burnchain_block_rx.recv() {
```

Comment:
I'm not sure why it is necessary to generate another burnchain block?  You should be able to generate and submit leader keys into the first burnchain block, and then insert block commits for them in any subsequent burnchain block.

Reply:
Sounds good, I can have some registered keys being part of the first burnchain block.

Reply:
My comment is actually confusing. By "another block" I meant block #1, the chain tip at this point is the genesis block.